Remove link between type id and value type

alla3 opened the pull request:
Remove the link between the KNOT_TYPE_ID and the KNOT_VALUE_TYPE
allowing the user to choose the most suitable type for its application.

alla3 committed:
Remove link between type id and value type

Remove the link between the KNOT_TYPE_ID and the KNOT_VALUE_TYPE
allowing the user to choose the most suitable type for its application.

netoax commented on src/knot_protocol.c:
```diff
@@ -172,21 +159,20 @@ int knot_type_id_is_generic(uint16_t type_id)
 				return 0;
 		} else if (knot_type_id_is_generic(type_id) == 0) {
-			if (generic_types[TYPE_ID_MASK(type_id)] == value_type)
+			if (unit == KNOT_UNIT_NOT_APPLICABLE)
```
Are generic values covering analogic values? If so, does it make sense to limit the unit to be not applicable?

alla3 replied:
Yes. I am not sure. I tried to follow the same pattern described in [KNoT Documentation](https://knot-devel.cesar.org.br/doc/thing/unit-type-value.html) and implemented in [KNoT BabelTower schema rules](https://github.com/CESARBR/knot-babeltower/blob/b4a9de80f482495740733a5a0f71daf43a54201e/pkg/thing/interactors/update_schema.go#L47). What do you think @vitbaq?

netoax replied:
@vitbaq what do you think?

k-cabral replied:
Because each unit is invariably tied to some other Type Id, I don't think it would make sense to allow other units. That being said, it's not clear to me what's the use case for Type Id Analog.

vitbaq replied:
I suggest changing the name from analog to generic.
KNOT_TYPE_ID_ANALOG -> KNOT_TYPE_ID_GENERIC

If you use an analogic sensor you can choose any type_id, including the generic type.
The generic type represents all other types that are not mapped by the protocol. And i think that the only unit that can be used is KNOT_UNIT_NOT_APPLICABLE. Or we can create a new unit: KNOT_UNIT_GENERIC

netoax replied:
It's now making sense for me. I think we can go with `KNOT_UNIT_NOT_APPLICABLE`.

alla3 replied:
I add a commit replacing the KNOT_TYPE_ID_ANALOG with KNOT_TYPE_ID_GENERIC and adding the KNOT_UNIT_GENERIC.

netoax reviewed (commented):
Nice patch.
I just made one remark.

alla3 committed:
Add support for new types in config

Add support for the int64, uint and uint64 types in config.

k-cabral commented on src/knot_protocol.c:
```diff
@@ -172,21 +159,20 @@ int knot_type_id_is_generic(uint16_t type_id)
 		/* Verify logic type IDs */
 		} else if (knot_type_id_is_logic(type_id) == 0) {
-			if (logic_types[TYPE_ID_MASK(type_id)] == value_type)
+			if (unit == KNOT_UNIT_NOT_APPLICABLE)
 				return 0;
 		} else if (knot_type_id_is_generic(type_id) == 0) {
-			if (generic_types[TYPE_ID_MASK(type_id)] == value_type)
+			if (unit == KNOT_UNIT_NOT_APPLICABLE)
 				return 0;
 		}
```
Could these checks be merged? They're doing the same thing.

alla3 replied:
Yes, I tried to keep the same code structure since it was representing different groups of type ids, but I can change it if you prefer.

ramonhpr commented on src/knot_protocol.c:
```diff
@@ -98,42 +98,29 @@
-	{KNOT_VALUE_TYPE_INT, UNIT_ENERGY_MASK		}
-};
-
-static const uint8_t logic_types[]  = {
-	KNOT_VALUE_TYPE_BOOL,			// KNOT_TYPE_ID_PRESENCE
-	KNOT_VALUE_TYPE_BOOL,			// KNOT_TYPE_ID_SWITCH
-	KNOT_VALUE_TYPE_RAW			// KNOT_TYPE_ID_COMMAND
-};
```
I know that I had talked to you in private @alla3 . But I'd like to bring what we discussed in the PR.
When we remove the type logic we will allow schemas with scenarios like for example temperatures using boolean values.
The question is if that is responsability to the user or the protocol.

vitbaq replied:
I think we can't tie the value type with the semantic, because there are several use cases that we were unable to map. So, I believe that we can leave the responsibility to the user.

netoax replied:
I agree with you @vitbaq, when talking about the responsibility. Also, the discussion could be summarized in this question: how strict will be the protocol defined by us?
IMO, it's a tradeoff between that and our ability to maintain this protocol and be responsible to cover a lot of scenarios.

alla3 committed:
Replace Type ID Analog with Type ID Generic

Replace KNOT_TYPE_ID_ANALOG with KNOT_TYPE_ID_GENERIC since analog
sensors can be registered using more descriptives types. Thus the
KNOT_TYPE_ID_GENERIC should be used with types not covered by the
KNoT Protocol. This patch also adds the unit KNOT_UNIT_GENERIC.

netoax reviewed (approved):
LGTM!
